### ISSUE TITLE
make[2]: *** No rule to make target '... test', needed by 'lib/libmlpack.so.3.4'.  Stop.

### ISSUE THREAD

**Issue by ttbek** [mlpack/mlpack#2937]
When building on Ubuntu 20.04, I eventually get the error: 
```
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_coding/sparse_coding.cpp.o
make[2]: *** No rule to make target '/home/name/Programming/ML_Programming/boost_1_76_0/test', needed by 'lib/libmlpack.so.3.4'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:2577: src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```
I can post the whole thing, but there is a lot of just compiler warning stuff (probably from Boost, about deprecation of this and that and unused parameters).
My CMake I thought it is finding what it needs, though it does warn about the new Boost (I also tried an older one with identical results other than the warning):
```
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- cotire 1.7.10 loaded.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Looking for sgemm_
-- Looking for sgemm_ - not found
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so  
-- Found Armadillo: /home/name/Programming/ML_Programming/armadillo-10.4.1/include (found suitable version "10.4.1", minimum required is "8.400.0") 
-- Found StbImage: /usr/include/stb/  
-- Found Ensmallen: /home/name/Programming/ML_Programming/ensmallen-2.16.2/include (found suitable version "2.16.2", minimum required is "2.10.0") 
CMake Warning at /usr/share/cmake-3.16/Modules/FindBoost.cmake:1161 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1283 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1921 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:442 (find_package)


CMake Warning at /usr/share/cmake-3.16/Modules/FindBoost.cmake:1161 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1283 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1921 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:442 (find_package)


-- Found Boost: /home/name/Programming/ML_Programming/boost_1_76_0 (found suitable version "1.76.0", minimum required is "1.58") found components: unit_test_framework serialization 
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Found Git: /usr/bin/git (found version "2.25.1") 
CMake Warning at CMakeLists.txt:585 (message):
  txt2man not found; man pages will not be generated.


-- Found Julia: /opt/julia-1.1.0-linux-x86_64/julia-1.1.0/bin/julia (found suitable version "1.1.0", minimum required is "0.7.0") 
-- Not building Markdown bindings.
-- Not building Python bindings.
-- Go not found (missing: GO_EXECUTABLE) (Required is at least version "1.11.0")
-- Gonum not found (missing: GONUM_VERSION_STRING) 
-- Not building Go bindings; the following modules are not available: 
    - Go
    - Gonum
-- R not found (missing: R_EXECUTABLE) (Required is at least version "4.0")
-- Could NOT find R_roxygen2 (missing: R_ROXYGEN2) 
-- Could NOT find R_Rcpp (missing: R_RCPP) 
-- Could NOT find R_RcppArmadillo (missing: R_RCPPARMADILLO) 
-- Could NOT find R_RcppEnsmallen (missing: R_RCPPENSMALLEN) 
-- Could NOT find R_BH (missing: R_BH) 
-- Could NOT find R_testthat (missing: R_TESTTHAT) 
-- Not building R bindings, Following modules are not available:
    - R
    - roxygen2
    - Rcpp
    - RcppArmadillo
    - RcppEnsmallen
    - BH
    - testthat
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- CXX target mlpack cotired.
-- CXX target mlpack_test cotired.
-- CXX target mlpack_catch_test cotired.
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/name/Programming/ML_Programming/mlpack-3.4.2/build
```
I specified the directories explicitly as I am installing these outside of my LD_LIBRARY_PATH, etc... 
```
cmake .. \
-DCMAKE_INSTALL_PREFIX:PATH=/home/name/Programming/ML_Programming/mlpack_compiled \
-DARMADILLO_INCLUDE_DIR:PATH=/home/name/Programming/ML_Programming/armadillo-10.4.1/include/ \
-DENSMALLEN_INCLUDE_DIR:PATH=/home/name/Programming/ML_Programming/ensmallen-2.16.2/include/ \
-DBOOST_ROOT:PATH=/home/name/Programming/ML_Programming/boost_1_76_0/ \
-DBoost_INCLUDE_DIR:PATH=/home/name/Programming/ML_Programming/boost_1_76_0/ \
-DBoost_LIBRARY_DIR:PATH=/home/name/Programming/ML_Programming/boost_1_76_0/libs/ \
-DBoost_UNIT_TEST_FRAMEWORK_LIBRARY_RELEASE:FILEPATH=/home/name/Programming/ML_Programming/boost_1_76_0/test/ \
-DBoost_SERIALIZATION_LIBRARY_RELEASE:FILEPATH=/home/name/Programming/ML_Programming/boost_1_76_0/serialization/ \
-DBoost_UNIT_TEST_FRAMEWORK_LIBRARY_DEBUG:FILEPATH=/home/name/Programming/ML_Programming/boost_1_76_0/test/ \
-DBoost_SERIALIZATION_LIBRARY_DEBUG:FILEPATH=/home/name/Programming/ML_Programming/boost_1_76_0/serialization/
```
Though some of those are likely redundant.  I was surprised it doesn't pick up the unit test framework or serialization when given the boost root, but having given everything explicitly the cmake seemed happy, but the make install is not.

**Comment by zoq** [comment]
The reason why cmake complains about boost, is that it doesn't know the version, unfortunately we have to update the CMake file for each new version that comes out. So either you update `CMakeLists.txt` manually and bring it to the current version:

https://github.com/mlpack/mlpack/blob/596859cb5bd7217a2f159097e1edb9d5a7f2ae83/CMakeLists.txt#L326-L347

or you build the git master branch. Let me know if that helps.

We are about to remove the boost dependencies, so that issue will be resolved soonish.

**Comment by heisenbuug** [comment]
@zoq Not sure, but maybe we should mention this issue in the documentation somewhere? I remember someone else was also asking about the same issue and I myself have faced the same.

**Comment by ttbek** [comment]
@zoq Sorry that I wasn't clear, I did see those issues so I also tried compiling with Boost 1.63.0 as well.  It removes the complaints, i.e. 
```
CMake Warning at /usr/share/cmake-3.16/Modules/FindBoost.cmake:1161 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1283 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:1921 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:442 (find_package)
```
at the cmake step, but the make install is almost (maybe entirely, some unused vars and deprecation complaints are maybe different, there is a lot of text) identical and gives the same error: 
```
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_coding/sparse_coding.cpp.o
make[2]: *** No rule to make target '/home/name/Programming/ML_Programming/boost_1_76_0/test', needed by 'lib/libmlpack.so.3.4'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:2577: src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```
but of course with "boost_1_63_0" in place of "boost_1_76_0."

**Comment by shrit** [comment]
@ttbek Do you have boost installed on your machine?

**Comment by ttbek** [comment]
@shrit Umm, what do you mean by "installed," it is a template-only library and I pointed to the appropriate locations explicitly.  Do I need to compile Boost as a static/shared lib and point to that instead?  I'm guessing I do... will see this "b2" and be back.

**Comment by rcurtin** [comment]
@ttbek awesome that you got it figured out!  We are (slowly) in the process of removing our boost dependency, so, hopefully when that is gone, strange linking issues like this should become a lot less prevalent. :smile:

**Comment by shrit** [comment]
@ttbek Boost is not a header library, and it is far from being header-only . Therefore, boost requires installation. 
On Linux mlpack build a shared library and executables, and links against a shared oned too, if you would like to change that, you can do this by specifying cmake option `cmake -DBUILD_SHARED_LIBS=OFF`, and then continue normally with `make`

**Comment by ttbek** [comment]
@shrit Yes, I see that now, I was browsing too quickly here:
https://www.boost.org/doc/libs/1_76_0/more/getting_started/unix-variants.html

And misinterpreted language like 

> 3   Header-Only Libraries
> 
> The first thing many people want to know is, “how do I build Boost?” The good news is that often, there's nothing to build.
> 
> Nothing to Build?
> 
> Most Boost libraries are header-only: they consist entirely of header files containing templates and inline functions, and require no separately-compiled library binaries or special treatment when linking.

In my haste, totally my fault.  Thank you very much for the cmake option, I do hope to be able to have static builds of my final programs, guess I need to recompile then, will actually use '-j' this time.

**Comment by ttbek** [comment]
I was going to post the following as an issue with the examples, but I've figured it out and just thought it might be helpful for anyone else being a noob like me that needed the above.  

Having successfully compiled my static mlpack, I tested a simple example and that worked, I then tried out another example ( https://github.com/mlpack/examples/blob/master/mnist_cnn/mnist_cnn.cpp ) and it was failing for me with a ton of output mentioning more than one lib.

At first I was trying to mess about with the make file, but I was having trouble getting it to put the .cpp file before the includes (and linkers are order sensitive so I thought that was maybe the issue) but I eventually managed to get the order correct and the linker error was still happening.  

It looked to be mostly about serialization and I figured out that commenting out 206 and 225 (the lines to do with saving) lets it compile fine, turns out only 206 needed commenting out, but I didn't know that at the time.  

My command for building was: 
```
g++ example.cpp -std=c++11 -DARMA_DONT_USE_WRAPPER \
-DARMA_USE_BLAS -DARMA_USE_LAPACK \
-I /home/name/Programming/ML_Programming/mlpack_compiled/include/ \
-I /home/name/Programming/ML_Programming/armadillo-10.4.1/include/ \
-I /home/name/Programming/ML_Programming/ensmallen-2.16.2/include/ \
-I /home/name/Programming/ML_Programming/boost_compiled/include/ \
-L /home/name/Programming/ML_Programming/mlpack_compiled/lib/ \
-L /home/name/Programming/ML_Programming/boost_compiled/lib/ \
-o example -fopenmp -lopenblas -llapack
```
It works if that line is commented
I wasn't sure at the time, did I need to include cereal somehow?  As far as I can tell cereal is listed as a requirement now but doesn't seem to be used in the build process yet (since Boost isn't totally kicked out yet I guess)?

Let's see if I have this right, ensmallen is header only.  Armadillo is header only when done this way with the option not to use the wrapper.  mlpack is statically compiled and the file libmlpack.a is in the respective lib folder.  Boost compiled both ways (static and shared), so the boost lib folder contains both .so and .a files.  For the moment openmp, openblas, and lapack are using the apt-get installed system libraries.  

The error is long, I can get all of it if needed, the last few lines are: 
```
/usr/bin/ld: /tmp/cc19lTkO.o: in function `void boost::archive::basic_text_oprimitive<std::ostream>::save_impl<boost::archive::class_id_type>(boost::archive::class_id_type const&, mpl_::bool_<false>&)':
example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13class_id_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13class_id_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0x7b): undefined reference to `boost::archive::archive_exception::archive_exception(boost::archive::archive_exception::exception_code, char const*, char const*)'
/usr/bin/ld: example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13class_id_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13class_id_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0xd6): undefined reference to `boost::archive::archive_exception::~archive_exception()'
/usr/bin/ld: /tmp/cc19lTkO.o: in function `void boost::archive::basic_text_oprimitive<std::ostream>::save_impl<boost::archive::class_id_reference_type>(boost::archive::class_id_reference_type const&, mpl_::bool_<false>&)':
example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_23class_id_reference_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_23class_id_reference_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0x7b): undefined reference to `boost::archive::archive_exception::archive_exception(boost::archive::archive_exception::exception_code, char const*, char const*)'
/usr/bin/ld: example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_23class_id_reference_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_23class_id_reference_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0xd6): undefined reference to `boost::archive::archive_exception::~archive_exception()'
/usr/bin/ld: /tmp/cc19lTkO.o: in function `void boost::archive::basic_text_oprimitive<std::ostream>::save_impl<boost::archive::tracking_type>(boost::archive::tracking_type const&, mpl_::bool_<false>&)':
example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13tracking_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13tracking_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0x7b): undefined reference to `boost::archive::archive_exception::archive_exception(boost::archive::archive_exception::exception_code, char const*, char const*)'
/usr/bin/ld: example.cpp:(.text._ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13tracking_typeEEEvRKT_RN4mpl_5bool_ILb0EEE[_ZN5boost7archive21basic_text_oprimitiveISoE9save_implINS0_13tracking_typeEEEvRKT_RN4mpl_5bool_ILb0EEE]+0xd8): undefined reference to `boost::archive::archive_exception::~archive_exception()'
/usr/bin/ld: /tmp/cc19lTkO.o: in function `void boost::archive::save_access::save_primitive<boost::archive::binary_oarchive, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(boost::archive::binary_oarchive&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
example.cpp:(.text._ZN5boost7archive11save_access14save_primitiveINS0_15binary_oarchiveENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRT_RKT0_[_ZN5boost7archive11save_access14save_primitiveINS0_15binary_oarchiveENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRT_RKT0_]+0x1c): undefined reference to `boost::archive::detail::basic_oarchive::end_preamble()'
/usr/bin/ld: example.cpp:(.text._ZN5boost7archive11save_access14save_primitiveINS0_15binary_oarchiveENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRT_RKT0_[_ZN5boost7archive11save_access14save_primitiveINS0_15binary_oarchiveENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRT_RKT0_]+0x33): undefined reference to `boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /tmp/cc19lTkO.o:(.data.rel.local.DW.ref._ZTIN5boost7archive17archive_exceptionE[DW.ref._ZTIN5boost7archive17archive_exceptionE]+0x0): undefined reference to `typeinfo for boost::archive::archive_exception'
collect2: error: ld returned 1 exit status
```
So, as it turns out, what I wasn't catching here is that static libraries don't auto-include their next order dependency per se.  So while linking the mlpack shared library would have pulled in the appropriate Boost lib, here I needed to do it manually.  So I needed to add in Boost Serialization, like so: 
<pre>
<code>
g++ example.cpp -std=c++11 -DARMA_DONT_USE_WRAPPER \
-I /home/name/Programming/ML_Programming/mlpack_compiled/include/ \
-I /home/name/Programming/ML_Programming/armadillo-10.4.1/include/ \
-I /home/name/Programming/ML_Programming/ensmallen-2.16.2/include/ \
-I /home/name/Programming/ML_Programming/boost_compiled/include/ \
-L /home/name/Programming/ML_Programming/mlpack_compiled/lib/ \
-L /home/name/Programming/ML_Programming/boost_compiled/lib/ \
-o example <b>-lboost_serialization</b> -fopenmp -lopenblas -lmlpack
</code>
</pre>

Is there a document mapping the Boost dependencies still in use (because there are a bunch of separate .a files from boost, it was a bit of a guess here for me, especially with there being another one related to serialization?  I guess the other dependencies are more obvious, though I'm not certain on the just how optional blas and lapack are, if I need to use these: "-DARMA_USE_BLAS and -DARMA_USE_LAPACK" or not in order to link the shared libs, those flags don't seem to change the executable size at all.  

Sorry to harp on in this thread, I just hope it might be helpful to someone in the interim before Boost is removed.

**Comment by shrit** [comment]
@ttbek the examples are modified according to the master branch of mlpack, not with version 3.4.2.
Therefore we have removed `-lboost_serialization` from the `Makefile`.
If you are not using mlpack in production, but only experimenting, it would be better (in my opinion) to use the `master` branch since it has removed all boost libraries that require linking.
If you use the `master` version, you need to have cereal installed on your system, so the compiler will find the headers. Or you can install the cereal header somewhere and point out the location the installation
Cereal replaced boost serialization since it does not require linking and require much less space than boost serialization.
Basically now you need to link only with openblas or Armadillo if you are using the `master` branch.
